### PR TITLE
Diagnose problematic uses of constructor/destructor attribute

### PR DESCRIPTION
**Pull request by AaronBallman**
Functions with these attributes will be automatically called before main() or after main() exits gracefully. In glibc environments, the constructor function is passed the same arguments as main(), so that signature is allowed. In all other environments, we require the function to accept no arguments and either return `void` or `int`. The functions must use the C calling convention. In C++ language modes, the functions cannot be a nonstatic member function, or a consteval function.

Additionally, these reuse the same priority logic as the init_priority attribute which explicitly reserved priorty values <= 100 or > 65535. So we now diagnose use of reserved priorities the same as we do for the init_priority attribute, but we downgrade the error to be a warning which defaults to an error to ease use for implementers like compiler-rt or libc.

### CODE REVIEW

**Comment by AaronBallman** [comment]
> I _think_ I made the correct changes in compiler-rt's CMake so that all of the compiler-rt projects can use reserved priorities, but I'd appreciate confirmation from @MaskRay on those changes.

Hmm, nope, it seems the test cases still fail (but the source files work).

**Comment by AaronBallman** [comment]
Ping, especially @MaskRay

**Comment by jhuber6** [comment]
I already changed the `init_fini_array_test` while I fixed an unrelated problem, so it should be fine now.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 1]
```diff
@@ -2352,26 +2352,126 @@ static void handleUnusedAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+
+static bool FunctionParamsAreMainLike(ASTContext &Context,
+                                      const FunctionDecl *FD) {
+  assert(FD->hasPrototype() && "expected the function to have a prototype");
+  const auto *FPT = FD->getType()->castAs<FunctionProtoType>();
+  QualType CharPP =
+      Context.getPointerType(Context.getPointerType(Context.CharTy));
+  QualType Expected[] = {Context.IntTy, CharPP, CharPP, CharPP};
+
+  for (unsigned I = 0; I < FPT->getNumParams(); ++I) {
```
Comment (nickdesaulniers) [inline clang/lib/Sema/SemaDeclAttr.cpp:2379]: I'm worries about the expression `Expected[I]` going out of bounds below. I think if you add `&& I < 4` to the loop condition that might be an additional safeguard.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 2]
```diff
@@ -2352,26 +2352,126 @@ static void handleUnusedAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+          (PT = Qs.strip(PT->getPointeeType())->getAs<PointerType>()) &&
+          Context.hasSameType(QualType(Qs.strip(PT->getPointeeType()), 0),
+                              Context.CharTy)) {
+        Qs.removeConst();
+        Mismatch = !Qs.empty();
+      }
+    }
+
+    if (Mismatch)
+      return false;
```
Comment (nickdesaulniers) [inline clang/lib/Sema/SemaDeclAttr.cpp:2403]: I think it's possible to express this without `Mismatch`:

```c++
if (!Context.hasSameUnqualifiedType(AT, Expected[I]) && Expected[I] == CharPP) {
  QualifierCollector Qs;
  ...
    if (!Qs.empty())
      return false;
```

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 1 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:2379]: Sure!

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 2 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:2403]: Good call, I've simplified.

**Review by nickdesaulniers — APPROVED** [review]

**Comment by AaronBallman** [comment]
I landed this and compiler-rt is still broken. I tried again to fix it up and wasn't able to find the right way to convince CMake to suppress the diagnostic. I have run out of time to put into trying to solve the build system issues (this was a drive-by set of fixes that have now eaten up more time than I can afford to spend on this), so abandoning this work. Anyone who wants to take it across the finish line is welcome to do so.

**Comment by Endilll** [comment]
Tagging @llvm/pr-subscribers-compiler-rt-sanitizer for visibility again, since this PR is abandoned.

**Comment by MaskRay** [comment]
(Was out of town...)

As mentioned at https://github.com/llvm/llvm-project/pull/67360#issuecomment-1736187374 , I think a default error is probably not appropriate. A priority value <= 100 may compete with a runtime `__attribute__((constructor(X)))` but more often it will work. A warning is fine.

**Comment by AaronBallman** [comment]
> (Was out of town...)
> 
> As mentioned at [#67360 (comment)](https://github.com/llvm/llvm-project/pull/67360#issuecomment-1736187374) , I think a default error is probably not appropriate. A priority value <= 100 may compete with a runtime `__attribute__((constructor(X)))` but more often it will work. A warning is fine. GCC defaults to a warning and the attribute form with a priority is an obscure feature anyway, therefore it doesn't buy us much to make it an error. You can make `> 65535` an error, though.

I disagree -- we decided ages ago with init_priority that using these reserved priorities should be an error that can be ignored for the few cases where the reservation does not apply. We should be designing the Ux for these features around the majority of users, not the majority of people who will use the feature. Otherwise, we should not pretend these are reserved values -- it's far too easy to ignore a warning IMO.

That said, I'm abandoning this work regardless. It was unplanned work I took on only because I noticed how deficient the implementation was regarding diagnostics when I went to add documentation for the attribute in the first place. A lot of our attributes from 2000-2012 or so are similarly problematic, unfortunately. So I wouldn't be surprised to see this come up again in the future.

**Comment by erichkeane** [comment]
> > (Was out of town...)
> > As mentioned at [#67360 (comment)](https://github.com/llvm/llvm-project/pull/67360#issuecomment-1736187374) , I think a default error is probably not appropriate. A priority value <= 100 may compete with a runtime `__attribute__((constructor(X)))` but more often it will work. A warning is fine. GCC defaults to a warning and the attribute form with a priority is an obscure feature anyway, therefore it doesn't buy us much to make it an error. You can make `> 65535` an error, though.
> 
> I disagree -- we decided ages ago with init_priority that using these reserved priorities should be an error that can be ignored for the few cases where the reservation does not apply. We should be designing the Ux for these features around the majority of users, not the majority of people who will be expert user of the feature. Otherwise, we should not pretend these are reserved values -- it's far too easy to ignore a warning IMO.
> 
> That said, I'm abandoning this work regardless. It was unplanned work I took on only because I noticed how deficient the implementation was regarding diagnostics when I went to add documentation for the attribute in the first place. A lot of our attributes from 2000-2012 or so are similarly problematic, unfortunately. So I wouldn't be surprised to see this come up again in the future.

First, it would be a shame if you abandon this over this bit, there is plenty of value to this patch, even as far as the refactoring (which makes some other folks' patches easier to review!).  While I agree with you that the error is appropriate, it doesn't seem practical at the moment thanks to GCC's implementation.  Abandoning this patch means that those will NEVER be an error, so at least having this patch in gives us an opportunity to make this an error 'someday'.

IMO, a warning/Warning-as-default-error are both appropriate (with preference towards the latter) 'for now', and we should fight the battle of making this a hard error another time.

**Comment by AaronBallman** [comment]
> > > (Was out of town...)
> > > As mentioned at [#67360 (comment)](https://github.com/llvm/llvm-project/pull/67360#issuecomment-1736187374) , I think a default error is probably not appropriate. A priority value <= 100 may compete with a runtime `__attribute__((constructor(X)))` but more often it will work. A warning is fine. GCC defaults to a warning and the attribute form with a priority is an obscure feature anyway, therefore it doesn't buy us much to make it an error. You can make `> 65535` an error, though.
> > 
> > 
> > I disagree -- we decided ages ago with init_priority that using these reserved priorities should be an error that can be ignored for the few cases where the reservation does not apply. We should be designing the Ux for these features around the majority of users, not the majority of people who will be expert user of the feature. Otherwise, we should not pretend these are reserved values -- it's far too easy to ignore a warning IMO.
> > That said, I'm abandoning this work regardless. It was unplanned work I took on only because I noticed how deficient the implementation was regarding diagnostics when I went to add documentation for the attribute in the first place. A lot of our attributes from 2000-2012 or so are similarly problematic, unfortunately. So I wouldn't be surprised to see this come up again in the future.
> 
> First, it would be a shame if you abandon this over this bit, there is plenty of value to this patch, even as far as the refactoring (which makes some other folks' patches easier to review!). While I agree with you that the error is appropriate, it doesn't seem practical at the moment thanks to GCC's implementation. Abandoning this patch means that those will NEVER be an error, so at least having this patch in gives us an opportunity to make this an error 'someday'.
> 
> IMO, a warning/Warning-as-default-error are both appropriate (with preference towards the latter) 'for now', and we should fight the battle of making this a hard error another time.

That doesn't solve the issue though -- we have `-Werror` bots for compiler-rt, so someone still needs to figure out how to beat CMake into shape even if we introduce it as a warning. I took several runs at the build system and was unsuccessful -- I can't repro locally on Windows and our precommit CI doesn't demonstrate the issue, so this requires making a number of speculative commits + watch the bots for fallout and that's where I ran out of time for the unexpected effort. I'm totally happy if someone wants to pick this up and run it across the finish line, warning-as-error or warning form. I personally think warning-as-error form is the correct approach though; "GCC doesn't err" isn't really compelling logic to me because GCC attributes (especially the older ones) have traditionally had underwhelming diagnostic behavior -- we should strive to do better when possible.

**Comment by erichkeane** [comment]
I missed the compiler-rt bit there, thats a real shame, this is a really valueable patch that I'd hope someone would take up.  While "GCC Doesnt Error" isn't particularly compelling and we SHOULD be better than that, I'd also prefer not throwing the baby out with the bath water.

That said, if you're unable to do the cmake work, I'm hopeful someone else can come along.  Is there a 'beginner commandeer' (ala-beginner-bug?) tag we could throw on this?

**Comment by Endilll** [comment]
If Aaron weren't able to fix this in several commits, I'm not sure it's a `good first issue`.

**Pull request merged**